File: running-app-backend/social/models/post_like.py
import uuid
from django.db import models
from django.db.models.signals import post_save, post_delete
from django.dispatch import receiver

class PostLike(models.Model):
    id = models.UUIDField(primary_key=True, default=uuid.uuid4, db_index=True)
    created_at = models.DateTimeField(auto_now_add=True)
    
    class Meta:
        abstract = True
        ordering = ['-created_at']
        
    def __str__(self):
        return f""

class EventPostLike(PostLike):
    RELATED_NAME = "event_post_likes"
    post = models.ForeignKey(
        "social.EventPost", related_name=RELATED_NAME, on_delete=models.CASCADE)
    user = models.ForeignKey(
        "account.Activity", related_name=RELATED_NAME, on_delete=models.CASCADE)

    def __str__(self):
        return f"{self.post} {self.user}"

class ClubPostLike(PostLike):
    RELATED_NAME = "club_post_likes"
    post = models.ForeignKey(
        "social.ClubPost", related_name=RELATED_NAME, on_delete=models.CASCADE)
    user = models.ForeignKey(
        "account.Activity", related_name=RELATED_NAME, on_delete=models.CASCADE)

    def __str__(self):
        return f"{self.post} - {self.user}"

class ActivityRecordPostLike(PostLike):
    RELATED_NAME = "act_rec_post_likes"
    post = models.ForeignKey(
        "activity.ActivityRecord", related_name=RELATED_NAME, on_delete=models.CASCADE)
    user = models.ForeignKey(
        "account.Activity", related_name=RELATED_NAME, on_delete=models.CASCADE)

    def __str__(self):
        return f"{self.post} - {self.user}"

# @receiver(post_save, sender=PostLike)
# def update_post_total_likes(sender, instance, created, **kwargs):
#     if created:
#         post = instance.post
#         post.total_likes += 1
#         post.save()

@receiver(post_save, sender=ClubPostLike)
def update_club_post_likes(sender, instance, created, **kwargs):
    if created:
        post = instance.post
        post.total_likes += 1
        post.save()

@receiver(post_save, sender=EventPostLike)
def update_event_post_likes(sender, instance, created, **kwargs):
    if created:
        post = instance.post
        post.total_likes += 1
        post.save()

@receiver(post_save, sender=ActivityRecordPostLike)
def update_activity_record_post_likes(sender, instance, created, **kwargs):
    if created:
        post = instance.post
        post.total_likes += 1
        post.save()

@receiver(post_delete, sender=ClubPostLike)
def update_club_post_likes_on_delete(sender, instance, **kwargs):
    post = instance.post
    post.total_likes -= 1
    post.save()

@receiver(post_delete, sender=EventPostLike)
def update_event_post_likes_on_delete(sender, instance, **kwargs):
    post = instance.post
    post.total_likes -= 1
    post.save()

@receiver(post_delete, sender=ActivityRecordPostLike)
def update_activity_record_post_likes_on_delete(sender, instance, **kwargs):
    post = instance.post
    post.total_likes -= 1
    post.save()

File: running_app/lib/models/social/post_like.dart
import 'package:running_app/models/account/activity.dart';
import 'package:running_app/models/account/like.dart';
import 'package:running_app/models/activity/activity_record.dart';
import 'package:running_app/models/social/post.dart';

class PostLike {
  final String? createdAt;

  PostLike({
    required this.createdAt,
  });

  PostLike.fromJson(Map<String, dynamic> json)
      : createdAt = json['created_at'];

  Map<String, dynamic> toJson() {
    return {
      'created_at': createdAt,
    };
  }

  @override
  String toString() {
    return '';
  }
}

class EventPostLike extends PostLike {
  final String? postId;
  final String? userId;

  EventPostLike({
    String? createdAt,
    required this.postId,
    required this.userId,
  }) : super(
      createdAt: createdAt
  );

  EventPostLike.fromJson(Map<String, dynamic> json)
      : postId = json['post_id'],
        userId = json['user_id'],
        super.fromJson(json);

  @override
  Map<String, dynamic> toJson() {
    return {
      'post_id': postId,
      'user_id': userId,
    };
  }

  @override
  String toString() {
    return '$postId - $userId';
  }
}

class ClubPostLike extends PostLike {
  final String? postId;
  final String? userId;

  ClubPostLike({
    String? createdAt,
    required this.postId,
    required this.userId,
  }) : super(createdAt: createdAt);

  ClubPostLike.fromJson(Map<String, dynamic> json)
      : postId = json['post_id'],
        userId = json['user_id'],
        super.fromJson(json);

  @override
  Map<String, dynamic> toJson() {
    return {
      'post_id': postId,
      'user_id': userId,
    };
  }

  @override
  String toString() {
    return '$postId - $userId';
  }
}

class ActivityRecordPostLike extends PostLike {
  final String? postId;
  final String? userId;

  ActivityRecordPostLike({
    String? createdAt,
    required this.postId,
    required this.userId,
  }) : super(createdAt: createdAt);

  ActivityRecordPostLike.fromJson(Map<String, dynamic> json)
      : postId = json['post_id'],
        userId = json['user_id'],
        super.fromJson(json);

  @override
  Map<String, dynamic> toJson() {
    return {
      'post_id': postId,
      'user_id': userId,
    };
  }

  @override
  String toString() {
    return '$postId - $userId';
  }
}

class CreatePostLike {
  final String? userId;
  final String? postId;

  CreatePostLike({
    required this.userId,
    required this.postId,
  });

  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'post_id': postId,
    };
  }

  @override
  String toString() {
    return "${toJson()}";
  }
}
